src/script/descriptor.cpp

```diff
@@ -1761,7 +1761,7 @@ struct KeyParser {
 std::vector<std::unique_ptr<DescriptorImpl>> ParseScript(uint32_t& key_exp_index, Span<const char>& sp, ParseScriptContext ctx, FlatSigningProvider& out, std::string& error)
 {
     using namespace script;
-
+    Assume(ctx != ParseScriptContext::P2WPKH);
```

Comment:
Seems like it might be clearer and more defensive to check that we're entering with one of the appropriate contexts rather than check that we aren't entering in one of the inappropriate contexts (albeit currently the only inappropriate one).

```diff
-    Assume(ctx != ParseScriptContext::P2WPKH);
+    Assume(ctx == ParseScriptContext::TOP || ctx == ParseScriptContext::P2SH || ctx == ParseScriptContext::P2WSH || ctx == ParseScriptContext::P2TR);

```

https://github.com/bitcoin/bitcoin/blob/fc7b21484703da606c5c69b23daee8c39506d90c/src/script/descriptor.cpp#L1406-L1412

Reply:
I'm not sure about this. I think `Assume(ctx != ParseScriptContext::P2WPKH)` is shorter and easier to understand.

Reply:
It's definitely shorter, for sure. A thought here is that if another context is added and it's not appropriate for the function, the check for allowed contexts would still work (conversely the !P2WPKH check would miss it). If another context is added and it is appropriate, then we're informed to add it (the assume would fail), which also serves as a reminder to ensure the function is updated appropriately.

Reply:
Yes, fair enough. I will address it.

Reply:
Done.